src/slurm_plugin/slurm_resources.py

```diff
@@ -0,0 +1,416 @@
+        return self.is_powering_up() and self.has_job()
+
+    def is_power_with_job(self):
+        """Dynamic nodes allocated a job but power up process has not started yet."""
+        return self.state == self.SLURM_SCONTROL_POWER_WITH_JOB_STATE
+
+    def is_running_job(self):
+        """Check if slurm node is running a job but not in configuring job state."""
+        return not self.is_powering_up() and self.has_job() and not self.is_power_with_job()
+
+    def is_resume_failed(self):
+        """Check if node resume timeout expires."""
+        return self.state == self.SLURM_SCONTROL_RESUME_FAILED_STATE
+
+    @abstractmethod
+    def is_state_healthy(self, terminate_drain_nodes, terminate_down_nodes):
+        """Check if a slurm node's scheduler state is considered healthy."""
+        pass
+
+    @abstractmethod
+    def is_bootstrap_failure(self):
+        """
+        Check if a slurm node has boostrap failure.
+
+        Here's the cases of bootstrap error we are checking:
+        Bootstrap error that causes instance to self terminate.
+        Bootstrap error that prevents instance from joining cluster but does not cause self termination.
+        """
+        pass
+
+    @abstractmethod
+    def is_healthy(self, terminate_drain_nodes, terminate_down_nodes):
+        """Check if a slurm node is considered healthy."""
+        pass
+
+    @abstractmethod
+    def is_powering_down_with_nodeaddr(self):
+        """Check if a slurm node is a powering down node with instance backing."""
+        pass
+
+    def is_backing_instance_valid(self):
+        """Check if a slurm node's addr is set, it points to a valid instance in EC2."""
+        if self.is_nodeaddr_set():
+            if not self.instance:
+                logger.warning("Node state check: no corresponding instance in EC2 for node %s", self)
+                return False
+        return True
+
+    @abstractmethod
+    def needs_reset_when_inactive(self):
+        """Check if the node need to be reset in node is inactive."""
+        pass
+
+    def __eq__(self, other):
+        """Compare 2 SlurmNode objects."""
+        if isinstance(other, SlurmNode):
+            return self.__dict__ == other.__dict__
+        return False
+
+    def __repr__(self):
+        attrs = ", ".join(["{key}={value}".format(key=key, value=repr(value)) for key, value in self.__dict__.items()])
+        return "{class_name}({attrs})".format(class_name=self.__class__.__name__, attrs=attrs)
+
+    def __str__(self):
+        return f"{self.name}({self.nodeaddr})"
+
+
+class StaticNode(SlurmNode):
+    def __init__(self, name, nodeaddr, nodehostname, state, partitions=None, instance=None):
+        super().__init__(name, nodeaddr, nodehostname, state, partitions, instance)
+        """Initialize slurm node with attributes."""
+        self.current_time = None
+        self.grace_time = None
+        self.static_nodes_in_replacement = {}
```

Comment:
Changed to is_being_replaced = True/False, move it to function _group_slurm_nodes, so it will be after the update of `_update_static_nodes_in_replacement`, didn't pass them to the function args because it is only need for static node but not for dynamic node, but _is_healthy, _is_state_healthy function are both for static nodes and dynamic nodes